src/webauthn/proto/raw_message.rs

```diff
@@ -45,14 +58,525 @@ pub struct Packed {
+                        let mut hash_name: Vec<u8> = vec![0, 11];
+                        let mut hasher = Sha256::new();
+                        hasher.update(self.pub_area.as_slice());
+                        let mut pub_area_hash = hasher.finalize().to_vec();
+                        hash_name.append(&mut pub_area_hash);
+                        hash_name
+                    }
+                    _ => return Err(Error::TpmError(TpmError::PubAreaHashUnknown(name_alg.into()))),
+                };
+
+                if hash_name != name.to_vec() {
+                    return Err(Error::TpmError(TpmError::AttestedNamePubAreaMismatch));
+                }
+            }
+            _ => return Err(Error::TpmError(TpmError::AttestationTypeInvalid)),
+        }
+
+        Ok(())
+    }
+
+    pub fn verify_extra_data(&self, auth_data: &[u8], client_data_hash: &[u8], extra_data: Option<Vec<u8>>) -> Result<(), Error> {
+        let mut buf = auth_data.to_vec();
+        buf.extend_from_slice(client_data_hash);
+
+        let att_to_be_signed = match self.alg.into() {
+            CoseAlgorithmIdentifier::RS1 => {
+                let mut hasher = Sha1::new();
+                hasher.update(buf.as_slice());
+                Ok(hasher.finalize_fixed().to_vec())
+            }
+            CoseAlgorithmIdentifier::RSA => {
+                let mut hasher = Sha256::new();
+                hasher.update(buf.as_slice());
+                Ok(hasher.finalize().to_vec())
+            }
+            _ => Err(Error::TpmError(TpmError::AttToBeSignedHashAlgorithmInvalid(self.alg))),
+        }?;
+
+        if let Some(extra_data) = extra_data {
+            if att_to_be_signed == extra_data {
+                return Ok(());
+            }
+        }
+
+        Err(Error::TpmError(TpmError::AttToBeSignedMismatch))
+    }
+
+    pub fn verify_signature(&self, cert: &[u8]) -> Result<(), Error> {
+        let (scheme, hashed) = match self.alg.into() {
+            CoseAlgorithmIdentifier::RS1 => (
+                Pkcs1v15Sign::new::<Sha1>(),
+                sha1::Sha1::digest(self.cert_info.as_slice()).as_slice().to_vec(),
+            ),
+            CoseAlgorithmIdentifier::RSA => (
+                Pkcs1v15Sign::new::<Sha256>(),
+                sha2::Sha256::digest(self.cert_info.as_slice()).as_slice().to_vec(),
+            ),
+            _ => return Err(Error::TpmError(TpmError::SignatureHashInvalid(self.alg))),
+        };
+
+        let public_key = RsaPublicKey::from_public_key_der(cert)
+            .map_err(|e| Error::Other(format!("verify_signature - Invalid certificate: {:?}", e)))?;
+        public_key
+            .verify(scheme, hashed.as_slice(), self.sig.as_slice())
+            .map_err(|_| Error::TpmError(TpmError::SignatureValidationFailed))
+    }
+
+    pub fn verify_public_key(&mut self, credential_pk: &CredentialPublicKey) -> Result<(), Error> {
+        let pub_area = PublicArea::from_vec(std::mem::take(&mut self.pub_area))?;
```

Comment:
This is destructive (because first time calling it will destroy the pub_area on self) but it doesn't have to be.
See the change in PublicArea::from_vec -> PublicArea::from_slice , and do this instead:
```suggestion
    pub fn verify_public_key(&self, credential_pk: &CredentialPublicKey) -> Result<(), Error> {
        let pub_area = PublicArea::from_slice(&self.pub_area)?;
```